src/fiber/fiber.ml

```diff
@@ -666,6 +666,40 @@ module Mvar = struct
         k ())
 end
 
+module Svar = struct
+  type 'a t =
+    { mutable current : 'a
+    ; mutable waiters : ('a K.t * ('a -> bool)) list
+    }
+
+  let read t = t.current
+
+  let wait t ~until k =
+    if until t.current then
+      k t.current
+    else
+      let wait = K.create k in
+      t.waiters <- (wait, until) :: t.waiters
+
+  let create current = { current; waiters = [] }
+
+  let write t a k =
+    t.current <- a;
+    let sleep, awake =
+      List.rev_partition_map t.waiters ~f:(fun (k, f) ->
+          if f t.current then
+            Right k
+          else
+            Left (k, f))
+    in
+    (match awake with
+    | [] -> ()
+    | awake ->
+      t.waiters <- List.rev sleep;
+      List.iter awake ~f:(fun k -> K.enqueue k a));
```

Comment:
BTW, just one thing about this implem: if between the time we enqueue the jobs and the time the readers are restarted there has been more writes, the readers will be awoken with a stale value.

Another option would be to not pass the value here, and just do `K.enqueue k ()`. Then `wait` would call itself recursively.

Reply:
When you think about it, every value will eventually be stale. So this is not a huge deal. But it still feels a tad odd to me that we would end up processing and displaying a state that we know is already stale.

Reply:
Another point that came up while discussing this offline with @snowleopard: the current implementation doesn't provide the following guarantee, that seems like a good guarantee:

```
when `Svar.wait` returns, the value it returns is the same as the current value of the ivar
```

The current implementation doesn't do that. Instead, it returns the first value that satisfied the predicate since `wait` was called. In a way this is a bit odd: when `wait` returns the current value of the `Svar` might not even satisfy the predicate anymore.

Reply:
I like this guarantee but it makes me think that the interface should perhaps be changed so that `wait` returns `unit t`. Then the caller is free to call the non-blocking `read` to get the actual value out. In this way, the invariant becomes a bit simpler: we just guarantee that calling `read` after `wait` returns a value that satisfies the predicate (no need to talk about these values being the same).

Reply:
> Another option would be to not pass the value here, and just do K.enqueue k (). Then wait would call itself recursively.

I tried doing this.

> I like this guarantee but it makes me think that the interface should be changed so that wait returns unit t. Then the caller is free to call the non-blocking read to get the actual value out

Yes, this is simpler and without any downsides.